File: src/main/java/com/weeth/domain/user/presentation/UserController.java
package com.weeth.domain.user.presentation;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import com.weeth.domain.user.application.dto.response.UserResponseDto;
import com.weeth.domain.user.application.dto.response.UserResponseDto.SummaryResponse;
import com.weeth.domain.user.application.dto.response.UserResponseDto.UserResponse;
import com.weeth.domain.user.application.exception.UserErrorCode;
import com.weeth.domain.user.application.usecase.UserManageUseCase;
import com.weeth.domain.user.application.usecase.UserUseCase;
import com.weeth.domain.user.domain.service.UserGetService;
import com.weeth.global.auth.annotation.CurrentUser;
import com.weeth.global.auth.jwt.application.dto.JwtDto;
import com.weeth.global.auth.jwt.exception.JwtErrorCode;
import com.weeth.global.common.exception.ApiErrorCodeExample;
import com.weeth.global.common.response.CommonResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Slice;
import org.springframework.web.bind.annotation.*;

import java.util.List;

import static com.weeth.domain.user.application.dto.request.UserRequestDto.*;
import static com.weeth.domain.user.application.dto.response.UserResponseDto.Response;
import static com.weeth.domain.user.application.dto.response.UserResponseDto.SocialLoginResponse;
import static com.weeth.domain.user.presentation.UserResponseCode.*;

@Tag(name = "USER", description = "사용자 API")
@RestController
@RequiredArgsConstructor
@RequestMapping("/api/v1/users")
@ApiErrorCodeExample({UserErrorCode.class, JwtErrorCode.class})
public class UserController {

    private final UserUseCase userUseCase;
    private final UserManageUseCase userManageUseCase;
    private final UserGetService userGetService;

    @PostMapping("/kakao/login")
    @Operation(summary = "카카오 소셜 로그인 API")
    public CommonResponse<SocialLoginResponse> login(@RequestBody @Valid Login dto) {
        SocialLoginResponse response = userUseCase.login(dto);
        return CommonResponse.success(SOCIAL_LOGIN_SUCCESS, response);
    }

    @PostMapping("/kakao/auth")
    @Operation(summary = "카카오 소셜 회원가입 전 요청 API (미사용 API)")
    public CommonResponse<UserResponseDto.SocialAuthResponse> beforeRegister(@RequestBody @Valid Login dto) {
        UserResponseDto.SocialAuthResponse response = userUseCase.authenticate(dto);
        return CommonResponse.success(SOCIAL_AUTH_SUCCESS, response);
    }

    @PostMapping("/apply")
    @Operation(summary = "동아리 지원 신청. 현재 사용하지 않으므로 회원가입 시 /kakao/register api로 요청 바람")
    public CommonResponse<Void> apply(@RequestBody @Valid SignUp dto) {
        userUseCase.apply(dto);
        return CommonResponse.success(USER_APPLY_SUCCESS);
    }

    @PostMapping("/kakao/register")
    @Operation(summary = "소셜 회원가입")
    public CommonResponse<Void> register(@RequestBody @Valid Register dto) {
        userUseCase.socialRegister(dto);
        return CommonResponse.success(USER_APPLY_SUCCESS);
    }

    @PatchMapping("/kakao/link")
    @Operation(summary = "카카오 소셜 로그인 연동")
    public CommonResponse<SocialLoginResponse> integrate(@RequestBody @Valid NormalLogin dto) {
        return CommonResponse.success(SOCIAL_INTEGRATE_SUCCESS, userUseCase.integrate(dto));
    }

    @PostMapping("/apple/login")
    @Operation(summary = "애플 소셜 로그인 API")
    public CommonResponse<SocialLoginResponse> appleLogin(@RequestBody @Valid Login dto) {
        SocialLoginResponse response = userUseCase.appleLogin(dto);
        return CommonResponse.success(SOCIAL_LOGIN_SUCCESS, response);
    }

    @PostMapping("/apple/register")
    @Operation(summary = "애플 소셜 회원가입 (dev 전용 - 바로 ACTIVE)")
    public CommonResponse<Void> appleRegister(@RequestBody @Valid Register dto) {
        userUseCase.appleRegister(dto);
        return CommonResponse.success(USER_APPLY_SUCCESS);
    }

    @GetMapping("/email")
    @Operation(summary = "이메일 중복 확인")
    public CommonResponse<Boolean> checkEmail(@RequestParam String email) {
        return CommonResponse.success(USER_EMAIL_CHECK_SUCCESS, userGetService.check(email));
    }

    @GetMapping("/all")
    @Operation(summary = "동아리 멤버 전체 조회(전체/기수별)")
    public CommonResponse<Slice<SummaryResponse>> findAllUser(@RequestParam("pageNumber") int pageNumber,
                                                              @RequestParam("pageSize") int pageSize,
                                                              @RequestParam(required = false) Integer cardinal) {
        return CommonResponse.success(USER_FIND_ALL_SUCCESS, userUseCase.findAllUser(pageNumber, pageSize, cardinal));
    }

    @GetMapping("/search")
    @Operation(summary = "동아리 멤버 검색")
    public CommonResponse<List<SummaryResponse>> searchUser(@RequestParam String keyword) {
        return CommonResponse.success(USER_FIND_BY_ID_SUCCESS, userUseCase.searchUser(keyword));
    }

    @GetMapping("/details")
    @Operation(summary = "특정 멤버 상세 조회")
    public CommonResponse<UserResponse> findUser(@RequestParam Long userId) {
        return CommonResponse.success(
                USER_DETAILS_SUCCESS, userUseCase.findUserDetails(userId)
        );
    }

    @GetMapping
    @Operation(summary = "내 정보 조회")
    public CommonResponse<Response> find(@Parameter(hidden = true) @CurrentUser Long userId) {
        return CommonResponse.success(USER_FIND_BY_ID_SUCCESS, userUseCase.find(userId));
    }

    @GetMapping("/info")
    @Operation(summary = "전역 내 정보 조회 API")
    public CommonResponse<UserResponseDto.UserInfo> findMyInfo(@Parameter(hidden = true) @CurrentUser Long userId) {
        return CommonResponse.success(USER_FIND_BY_ID_SUCCESS, userUseCase.findUserInfo(userId));
    }

    @PatchMapping
    @Operation(summary = "내 정보 수정")
    public CommonResponse<Void> update(@RequestBody @Valid Update dto, @Parameter(hidden = true) @CurrentUser Long userId) {
        userUseCase.update(dto, userId);
        return CommonResponse.success(USER_UPDATE_SUCCESS);
    }

    @DeleteMapping
    @Operation(summary = "동아리 탈퇴")
    public CommonResponse<Void> leave(@Parameter(hidden = true) @CurrentUser Long userId) {
        userManageUseCase.leave(userId);
        return CommonResponse.success(USER_LEAVE_SUCCESS);
    }

    @PostMapping("/refresh")
    @Operation(summary = "JWT 토큰 재발급 API")
    public CommonResponse<JwtDto> refresh(@Parameter(hidden = true) @RequestHeader("Authorization_refresh") String refreshToken) {
        return CommonResponse.success(JWT_REFRESH_SUCCESS, userUseCase.refresh(refreshToken));
    }
}

File: src/main/kotlin/com/weeth/domain/user/presentation/UserController.kt
package com.weeth.domain.user.presentation

import com.weeth.domain.user.application.dto.request.AgreeTermsRequest
import com.weeth.domain.user.application.dto.request.AssignClubProfileRequest
import com.weeth.domain.user.application.dto.request.CreateInquiryRequest
import com.weeth.domain.user.application.dto.request.CreateMultiProfileRequest
import com.weeth.domain.user.application.dto.request.SocialLoginRequest
import com.weeth.domain.user.application.dto.request.UpdateMultiProfileRequest
import com.weeth.domain.user.application.dto.request.UpdateUserProfileRequest
import com.weeth.domain.user.application.dto.response.SocialLoginResponse
import com.weeth.domain.user.application.dto.response.UserMyPageResponse
import com.weeth.domain.user.application.dto.response.UserProfileResponse
import com.weeth.domain.user.application.dto.response.UserProfilesResponse
import com.weeth.domain.user.application.exception.UserErrorCode
import com.weeth.domain.user.application.usecase.command.AgreeTermsUseCase
import com.weeth.domain.user.application.usecase.command.AuthUserUseCase
import com.weeth.domain.user.application.usecase.command.CreateInquiryUseCase
import com.weeth.domain.user.application.usecase.command.LeaveUserUseCase
import com.weeth.domain.user.application.usecase.command.ManageUserProfileUseCase
import com.weeth.domain.user.application.usecase.command.SocialLoginUseCase
import com.weeth.domain.user.application.usecase.command.UpdateUserProfileUseCase
import com.weeth.domain.user.application.usecase.query.GetUserMyPageQueryService
import com.weeth.domain.user.application.usecase.query.GetUserProfileQueryService
import com.weeth.global.auth.annotation.CurrentUser
import com.weeth.global.auth.jwt.application.dto.JwtDto
import com.weeth.global.auth.jwt.application.exception.JwtErrorCode
import com.weeth.global.auth.jwt.application.service.TokenCookieProvider
import com.weeth.global.common.exception.ApiErrorCodeExample
import com.weeth.global.common.response.CommonResponse
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.Parameter
import io.swagger.v3.oas.annotations.security.SecurityRequirements
import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.servlet.http.HttpServletRequest
import jakarta.validation.Valid
import org.springframework.http.HttpHeaders
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PatchMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@Tag(name = "USER", description = "사용자 API")
@RestController
@RequestMapping("/api/v4/users")
@ApiErrorCodeExample(UserErrorCode::class, JwtErrorCode::class)
class UserController(
    private val authUserUseCase: AuthUserUseCase,
    private val socialLoginUseCase: SocialLoginUseCase,
    private val updateUserProfileUseCase: UpdateUserProfileUseCase,
    private val agreeTermsUseCase: AgreeTermsUseCase,
    private val createInquiryUseCase: CreateInquiryUseCase,
    private val leaveUserUseCase: LeaveUserUseCase,
    private val manageUserProfileUseCase: ManageUserProfileUseCase,
    private val getUserProfileQueryService: GetUserProfileQueryService,
    private val getUserMyPageQueryService: GetUserMyPageQueryService,
    private val tokenCookieProvider: TokenCookieProvider,
) {
    @PostMapping("/social/kakao")
    @Operation(summary = "카카오 소셜 로그인(auth code flow)")
    @SecurityRequirements
    fun socialLoginByKakao(
        @RequestBody @Valid request: SocialLoginRequest,
    ): ResponseEntity<CommonResponse<SocialLoginResponse>> {
        val response = socialLoginUseCase.socialLoginByKakao(request)
        return buildTokenResponse(
            CommonResponse.success(UserResponseCode.SOCIAL_LOGIN_SUCCESS, response),
            response.accessToken,
            response.refreshToken,
        )
    }

    @PostMapping("/social/apple")
    @Operation(summary = "애플 소셜 로그인(auth code flow)")
    @SecurityRequirements
    fun socialLoginByApple(
        @RequestBody @Valid request: SocialLoginRequest,
    ): ResponseEntity<CommonResponse<SocialLoginResponse>> {
        val response = socialLoginUseCase.socialLoginByApple(request)
        return buildTokenResponse(
            CommonResponse.success(UserResponseCode.SOCIAL_LOGIN_SUCCESS, response),
            response.accessToken,
            response.refreshToken,
        )
    }

    @PostMapping("/social/refresh")
    @Operation(summary = "토큰 재발급", description = "쿠키를 사용해 토큰을 재발급합니다.")
    @SecurityRequirements
    fun refreshToken(request: HttpServletRequest): ResponseEntity<CommonResponse<JwtDto>> {
        val jwtDto = authUserUseCase.refreshToken(request)
        return buildTokenResponse(
            CommonResponse.success(UserResponseCode.JWT_REFRESH_SUCCESS, jwtDto),
            jwtDto.accessToken,
            jwtDto.refreshToken,
        )
    }

    @PostMapping("/terms")
    @Operation(summary = "약관 동의")
    fun agreeTerms(
        @RequestBody @Valid request: AgreeTermsRequest,
        @Parameter(hidden = true) @CurrentUser userId: Long,
    ): ResponseEntity<CommonResponse<JwtDto>> {
        val jwtDto = agreeTermsUseCase.execute(userId, request)
        return buildTokenResponse(
            CommonResponse.success(UserResponseCode.USER_TERMS_AGREE_SUCCESS, jwtDto),
            jwtDto.accessToken,
            jwtDto.refreshToken,
        )
    }

    @PatchMapping
    @Operation(summary = "내 정보 수정")
    fun update(
        @RequestBody @Valid request: UpdateUserProfileRequest,
        @Parameter(hidden = true) @CurrentUser userId: Long,
    ): CommonResponse<Void> {
        updateUserProfileUseCase.updateProfile(request, userId)
        return CommonResponse.success(UserResponseCode.USER_UPDATE_SUCCESS)
    }

    @DeleteMapping("/me")
    @Operation(summary = "위드 탈퇴")
    fun leave(
        @Parameter(hidden = true) @CurrentUser userId: Long,
    ): ResponseEntity<CommonResponse<Void>> {
        leaveUserUseCase.execute(userId)
        return buildExpiredTokenResponse(CommonResponse.success(UserResponseCode.USER_LEFT_SUCCESS))
    }

    @PostMapping("/inquiries")
    @Operation(summary = "문의하기")
    @SecurityRequirements
    fun createInquiry(
        @RequestBody @Valid request: CreateInquiryRequest,
    ): CommonResponse<Void> {
        createInquiryUseCase.execute(request)
        return CommonResponse.success(UserResponseCode.INQUIRY_SEND_SUCCESS)
    }

    @PostMapping("/me/profiles")
    @Operation(summary = "멀티프로필 생성")
    fun createUserProfile(
        @RequestBody @Valid request: CreateMultiProfileRequest,
        @Parameter(hidden = true) @CurrentUser userId: Long,
    ): CommonResponse<UserProfileResponse> {
        val response = manageUserProfileUseCase.create(userId, request)
        return CommonResponse.success(UserResponseCode.USER_PROFILE_CREATED_SUCCESS, response)
    }

    @GetMapping("/me/profiles")
    @Operation(summary = "멀티프로필 목록 조회")
    fun getUserProfiles(
        @Parameter(hidden = true) @CurrentUser userId: Long,
    ): CommonResponse<UserProfilesResponse> {
        val response = getUserProfileQueryService.findAll(userId)
        return CommonResponse.success(UserResponseCode.USER_PROFILE_FIND_ALL_SUCCESS, response)
    }

    @GetMapping("/me/profiles/{profileId}")
    @Operation(summary = "멀티프로필 단건 조회")
    fun getUserProfile(
        @PathVariable profileId: Long,
        @Parameter(hidden = true) @CurrentUser userId: Long,
    ): CommonResponse<UserProfileResponse> {
        val response = getUserProfileQueryService.find(userId, profileId)
        return CommonResponse.success(UserResponseCode.USER_PROFILE_FIND_SUCCESS, response)
    }

    @PatchMapping("/me/profiles/{profileId}")
    @Operation(summary = "멀티프로필 수정")
    fun updateUserProfile(
        @PathVariable profileId: Long,
        @RequestBody @Valid request: UpdateMultiProfileRequest,
        @Parameter(hidden = true) @CurrentUser userId: Long,
    ): CommonResponse<UserProfileResponse> {
        val response = manageUserProfileUseCase.update(userId, profileId, request)
        return CommonResponse.success(UserResponseCode.USER_PROFILE_UPDATED_SUCCESS, response)
    }

    @DeleteMapping("/me/profiles/{profileId}")
    @Operation(summary = "멀티프로필 삭제")
    fun deleteUserProfile(
        @PathVariable profileId: Long,
        @Parameter(hidden = true) @CurrentUser userId: Long,
    ): CommonResponse<Void> {
        manageUserProfileUseCase.delete(userId, profileId)
        return CommonResponse.success(UserResponseCode.USER_PROFILE_DELETED_SUCCESS)
    }

    @PatchMapping("/me/club-profile-assignments")
    @Operation(summary = "동아리별 사용 프로필 변경")
    fun assignClubProfiles(
        @RequestBody @Valid request: AssignClubProfileRequest,
        @Parameter(hidden = true) @CurrentUser userId: Long,
    ): CommonResponse<Void> {
        manageUserProfileUseCase.assignClubProfiles(userId, request)
        return CommonResponse.success(UserResponseCode.USER_PROFILE_ASSIGNMENT_UPDATED_SUCCESS)
    }

    @GetMapping("/me/mypage")
    @Operation(summary = "마이페이지 요약 조회")
    fun getMyPage(
        @Parameter(hidden = true) @CurrentUser userId: Long,
    ): CommonResponse<UserMyPageResponse> {
        val response = getUserMyPageQueryService.getMyPage(userId)
        return CommonResponse.success(UserResponseCode.USER_MY_PAGE_FIND_SUCCESS, response)
    }

    private fun <T> buildTokenResponse(
        body: CommonResponse<T>,
        accessToken: String,
        refreshToken: String,
    ): ResponseEntity<CommonResponse<T>> =
        ResponseEntity
            .ok()
            .header(HttpHeaders.SET_COOKIE, tokenCookieProvider.createAccessTokenCookie(accessToken).toString())
            .header(HttpHeaders.SET_COOKIE, tokenCookieProvider.createRefreshTokenCookie(refreshToken).toString())
            .body(body)

    private fun buildExpiredTokenResponse(body: CommonResponse<Void>): ResponseEntity<CommonResponse<Void>> =
        ResponseEntity
            .ok()
            .header(HttpHeaders.SET_COOKIE, tokenCookieProvider.expireAccessTokenCookie().toString())
            .header(HttpHeaders.SET_COOKIE, tokenCookieProvider.expireRefreshTokenCookie().toString())
            .body(body)
}
